library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -465,24 +521,32 @@ applyEquation term rule = fmap (either id Success) $ runExceptT $ do
             -- is violated
             checkConcreteness rule.attributes.concreteness subst
 
-            -- check conditions, using substitution (will call back
-            -- into the simplifier! -> import loop)
-            let newConstraints =
+            -- check required conditions, using substitution
+            let required =
                     concatMap (splitBoolPredicates . substituteInPredicate subst) $
                         rule.requires
-            unclearConditions' <- runMaybeT $ catMaybes <$> mapM checkConstraint newConstraints
+            unclearConditions' <- runMaybeT $ catMaybes <$> mapM checkConstraint required
 
             case unclearConditions' of
                 Nothing -> throwE ConditionFalse
                 Just unclearConditions ->
                     if not $ null unclearConditions
                         then throwE IndeterminateCondition
                         else do
-                            let rewritten =
-                                    substituteInTerm subst rule.rhs
-                            -- NB no new constraints, as they have been checked to be `Top`
-                            -- FIXME what about symbolic constraints here?
-                            pure rewritten
+                            -- check ensured conditions, filter any
+                            -- true ones, prune if any is false
+                            let ensured =
+                                    concatMap (splitBoolPredicates . substituteInPredicate subst) $
+                                        rule.ensures
+                            mbEnsuredConditions <-
+                                runMaybeT $ catMaybes <$> mapM checkConstraint ensured
+                            maybe
+                                (throwE $ EnsuresFalse ensured)
+                                -- throws if an ensured condition found to be false
+                                (lift . pushConstraints)
+                                -- pushes new ensured conditions and return result
+                                mbEnsuredConditions
```

Comment:
Is there a good reason to not do explicit pattern-matching on `mbEnsuredConditions`? I think `maybe` is generally harder to read than an explicit case statement.

```suggestion
                            case mbEnsuredConditions of
                                -- throws if an ensured condition found to be false
                                Nothing -> throwE $ EnsuresFalse ensured
                                -- pushes new ensured conditions and return result
                                Just conditions -> lift . pushConstraints $ conditions
```